go/gas-oracle/oracle/gas_price_oracle.go

```diff
@@ -0,0 +1,216 @@
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// errInvalidSigningKey represents the error when the signing key used
+// is not the Owner of the contract and therefore cannot update the gasprice
+var errInvalidSigningKey = errors.New("invalid signing key")
+
+// errNoChainID represents the error when the chain id is not provided
+// and it cannot be remotely fetched
+var errNoChainID = errors.New("no chain id provided")
+
+// errNoPrivateKey represents the error when the private key is not provided to
+// the application
+var errNoPrivateKey = errors.New("no private key provided")
+
+// GasPriceOracle manages a hot key that can update the L2 Gas Price
+type GasPriceOracle struct {
+	chainID   *big.Int
+	ctx       context.Context
+	stop      chan struct{}
+	contract  *bindings.GasPriceOracle
+	backend   DeployContractBackend
+	gasPricer *gasprices.GasPricer
+	config    *Config
+}
+
+// Start runs the GasPriceOracle
+func (g *GasPriceOracle) Start() error {
+	if g.config.chainID == nil {
+		return errNoChainID
+	}
+	if g.config.privateKey == nil {
+		return errNoPrivateKey
+	}
+
+	address := crypto.PubkeyToAddress(g.config.privateKey.PublicKey)
+	log.Info("Starting Gas Price Oracle", "chain-id", g.chainID, "address", address.Hex())
+
+	// Fetch the owner of the contract to check that the local key matches
+	// the owner of the contract. If it doesn't match then nothing can be
+	// accomplished.
+	owner, err := g.contract.Owner(&bind.CallOpts{
+		Context: g.ctx,
+	})
+	if err != nil {
+		return err
+	}
+
+	if address != owner {
+		log.Error("Signing key does not match contract owner", "signer", address.Hex(), "owner", owner.Hex())
+		return errInvalidSigningKey
+	}
+
+	// TODO: Errors in this goroutine should write to an error channel
+	// and be handled externally
+	go g.Loop()
+
+	return nil
+}
+
+func (g *GasPriceOracle) Stop() {
+	close(g.stop)
+}
+
+func (g *GasPriceOracle) Wait() {
+	<-g.stop
+}
+
+func (g *GasPriceOracle) Loop() {
+	tip, err := g.backend.HeaderByNumber(g.ctx, nil)
+	if err != nil {
+		log.Crit("Cannot fetch tip", "message", err)
+	}
+	// Start at the tip
+	epochStartBlockNumber := tip.Number.Uint64()
+	// getLatestBlockNumberFn is used by the GasPriceUpdater
+	// to get the latest block number
+	getLatestBlockNumberFn := wrapGetLatestBlockNumberFn(g.backend)
+	// updateL2GasPriceFn is used by the GasPriceUpdater to
+	// update the gas price
+	updateL2GasPriceFn, err := wrapUpdateL2GasPriceFn(g.backend, g.config)
+	if err != nil {
+		log.Crit("error", "message", err)
+	}
+
+	log.Info("Creating GasPriceUpdater", "epochStartBlockNumber", epochStartBlockNumber,
+		"averageBlockGasLimitPerEpoch", g.config.averageBlockGasLimitPerEpoch,
+		"epochLengthSeconds", g.config.epochLengthSeconds)
+
+	gasPriceUpdater, err := gasprices.NewGasPriceUpdater(
```

Comment:
An extension of my stylistic nit above -- Could consider adding a `gasPriceUpdater` to the `gas price oracle` & setting it at the same point that you set the `gasPricer` ( https://github.com/ethereum-optimism/optimism/pull/1103/files#diff-053eea5852d311aba5a88a1d44bc04f16b9ce3b05305c266c9fd345901eea940R181 )

Reply:
Ah did you set this here to make sure the `epochStartBlockNumber := tip.Number.Uint64()` is set to the tip as closely to when the loop starts as possible?

Reply:
Yes that is the reason as to why I create it here, I also didn't really see a reason to create a global gas price updater that is on the `GasOracle`